Implementing ViT for ADNI Dataset Classification - 47028643

Fhall21 opened the pull request:
## Pull Request: Implementing ViT for ADNI Dataset Classification

### Description
For this pull request, I am submitting my solution to the challenge of implementing a Visual Transformer (ViT) model to classify ADNI dataset. In the end I achieved an accuracy of 66.94% on the test dataset.

### Files and Folders Overview

Main files:
- **modules.py**: This file contains the code for the ViT model architecture and implementation.
- **dataset.py**: Code for load the dataset, as well as calculate the mean and standard deviation of the dataset.
- **train.py**: The primary script for training the model on the ADNI dataset, it is used to train, validate, test and save the model.
- **predict.py**: Script for evaluating the model's performance on and producing visuals from the test data.
- **driver.py**: This file is used to run the training and prediction scripts.

Relating to the README:
- **README.md**: A comprehensive README.md providing information about the model, the training process, and results.
- **/train_val_loss_curves/**: This folder contains the train and validation loss curves from hyperparameter tuning which is refered to within a table within the README.md
- **/example_train_images/**: This folder contains an example of the images used for training the model - also referred to within the README.md
- **/saved_models/train_visuals/45**: This folder path was the folder path used when saving model outputs, in this case the output visualisations are saved and the model file itself has been removed.
- **training_and_validation_loss_45.png**: This file contains the training and validation loss curves from the final training run of the model.
- **ViT-architecture.png**: This file contains an image of the ViT architecture used for this project.

### Note on Final Results
Overall, despite large hyperparameter tuning the model was not able to get much higher than 66%. In future, other variants of the ViT may be considered so as to gain such, but for the given architecture and hyperparameters, as evident by the testing table in the README.md, this was the best result I was able to achieve. Many of the other models tried tended to overfit and this was the best model when considering such as well as computational time (which realistically is a key resource but was not a key part of this challenge).

### Author

Felix Hall
47028643

Thank you for considering this pull request.

Commit:
wrote first skeleton version of ViT model

Commit:
first attempt at making the custom dataset

Commit:
wrote quick first draft of the training code; including splitting val and train dataset

Commit:
fixed dataset path

Commit:
first working test! Starting fine tuning, getting 66%.

Fhall21 committed:
fixed variable naming typo in Attention class

Fhall21 committed:
fixed bug in eval due to attention data returned from model being as a list

Commit:
tried adding attention visualisation, and testing out other optimisers (AdamW, Radam and SGD)

Fhall21 committed:
changed initialisation of model so as to change its architecture upon call - particularly its depth, n_heads and mlp ratio

Fhall21 committed:
adding in the visualisation export in the train file as well

Fhall21 committed:
changing the batch size and trying smaller models

Fhall21 committed:
playing around with embedding amount, learning rate and model properties

Fhall21 committed:
trying dropout probability with previous best model as that should improve overfitting

Fhall21 committed:
got high accuracy with a much larger model with no scheduler and smaller learning rate

Fhall21 committed:
now that we've got a compeltely different model, I wanted to try out AdamW again; small lr and large model

Fhall21 committed:
Testing small model with std and mean transformations

Fhall21 committed:
Finished more testing on a range of larger and smaller models and cannot get better than ~66%

Fhall21 committed:
moved files into recognition folder

Fhall21 committed:
moved to new branch

Fhall21 committed:
Merge branch 'topic-recognition' of https://github.com/Fhall21/PatternAnalysis-2023 into topic-recognition
merging remote topic branch with local topic branch

Fhall21 committed:
moved files into more appropriate folder - with name and student number

Fhall21 committed:
Wrote up teh README! Added in images, renamed files, and added in driver file - preparing for submission.

Fhall21 committed:
added header blocks to the main code files - brief explanations of what the file contains

Fhall21 committed:
re-worked some of the README.md wording to make it more legible and understandable

Fhall21 committed:
fixed a typo in the embed_dim hyperparameter! It was 76256 before - shocking

aryaman-sh commented:
**This is an initial inspection, no action is required at this point**

**Difficulty**
- Hard
- Task 6

**Readme**
- Excellent

**Commit messages**
- good

**Code**
- Uses vision transformer
- Good structure
- Good comments

**Functionality/Performance**
- accuracy 66.8%

**General comments**
- Good hyperparameter tuning
- No changes required

shakes76 commented:
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good

shakes76 commented:
Cannot merge because repo README deleted. Please remove DSTORE files as well. Does not affect grade only the merge.

wangzhaomxy commented:
No feedback attempt and no feedback marks lost.

wangzhaomxy commented:
No feedback attempt and no feedback marks lost.

Fhall21 committed:
Adding back repo README.mc

Fhall21 committed:
Delete recognition/.DS_Store as it was preventing the merge request

Fhall21 committed:
Delete other DS_STORE file preventing merge request: recognition/ViT-challenge-6-47028643/.DS_Store

Fhall21 committed:
Delete 3rd DS_Store file preventing merge request: recognition/ViT-challenge-6-47028643/train_val_loss_curves/.DS_Store

Fhall21 committed:
Delete DS_STORE file: recognition/ViT-challenge-6-47028643/saved_models/train_visuals/.DS_Store

Since it was preventing merge request - working through folder by folder; this is folder saved_models/train_visuals/45.

Fhall21 committed:
Delete DS_STORE file: recognition/ViT-challenge-6-47028643/saved_models/.DS_Store

Since it was preventing merge request - working through folder by folder; this is folder saved_models/train_visuals.

Fhall21 commented:
I have added the repo README (root folder README file which I assumed was the one you were referring to), as well as deleted the DSTORE files from each folder.

wangzhaomxy commented:
Feedback attempt and no feedback marks lost.
